### PR TITLE
[omim] Replaced boost::optional with std::optional.

### PR DESCRIPTION
**Pull request by mpimenov**
Since C++17, optional is part of the C++ Standard Library.

### CODE REVIEW

**Comment by mpimenov** [comment]
JTALL

**Comment by gmoryes** [comment]
https://www.bfilipek.com/2018/05/using-optional.html

**Comment by gmoryes** [comment]
```cpp
#include <iostream>
#include <optional>

class SomeClass
{
public:
  SomeClass(int x) : m_value(x) { std::cout << "SomeClass(" << x << ")" << std::endl; }
  SomeClass(SomeClass const & s) : m_value(s.m_value) { std::cout << "Copy" << std::endl; }
  SomeClass(SomeClass && s) : m_value(std::move(s.m_value)) { std::cout << "Move" << std::endl; }
  SomeClass & operator=(SomeClass const & s) { m_value = s.m_value; std::cout << "=(Copy)" << std::endl; return *this; }
  SomeClass & operator=(SomeClass && s) { m_value = std::move(s.m_value); std::cout << "=(Move)" << std::endl; return *this; }
  ~SomeClass() { std::cout << "~SomeClass()" << std::endl; }

private:
  int m_value;
};

int main()
{
  /*
    std::optional<SomeClass> op = SomeClass(123);
  
    SomeClass(123)
    Move
    ~SomeClass()
    ~SomeClass()
  */

  /* 
    auto op = std::make_optional<SomeClass>(321);
    
    SomeClass(123) 
    ~SomeClass()
  */
  
  /* 
    auto op = std::optional<SomeClass>(std::in_place, 123);
  
    SomeClass(123)  
    ~SomeClass()
  */

  /*  
    std::optional<SomeClass> op(123);

    SomeClass(123)
    ~SomeClass()
  */

  /*
    std::optional<SomeClass> op(123);
    op = 321; // or SomeClass(321) or std::optional<SomeClass>(321)

    SomeClass(123)
    SomeClass(321)
    =(Move)
    ~SomeClass()
    ~SomeClass()
  */
  
  /*
    std::optional<SomeClass> op(123);
    op.emplace(321);

    SomeClass(123)
    ~SomeClass()
    SomeClass(321)
    ~SomeClass()
  */
  
  return 0;
}
```

File: drape/vulkan/vulkan_base_context.cpp [hunk 1]
```diff
@@ -766,19 +766,19 @@ ref_ptr<VulkanStagingBuffer> VulkanBaseContext::GetDefaultStagingBuffer() const
   
 void VulkanBaseContext::RecreateSwapchain()
 {
-  CHECK(m_surface.is_initialized(), ());
-  CHECK(m_surfaceFormat.is_initialized(), ());
+  CHECK(m_surface.has_value(), ());
```
Comment (gmoryes) [inline drape/vulkan/vulkan_base_context.cpp]:
```cpp
CHECK(m_surface, ());
CHECK(m_surfaceFormat, ());
```

Comment (gmoryes) [reply drape/vulkan/vulkan_base_context.cpp]: ?

File: drape_frontend/frontend_renderer.cpp [hunk 2]
```diff
@@ -1320,10 +1320,11 @@ void FrontendRenderer::ProcessSelection(ref_ptr<SelectObjectMessage> msg)
   if (msg->IsDismiss())
   {
     m_selectionShape->Hide();
-    if (!m_myPositionController->IsModeChangeViewport() && m_selectionTrackInfo.is_initialized())
+    if (!m_myPositionController->IsModeChangeViewport() && m_selectionTrackInfo.has_value())
```
Comment (gmoryes) [inline drape_frontend/frontend_renderer.cpp]: without has_value?

File: drape/vulkan/vulkan_base_context.cpp (hunk 1 above)
Comment (gmoryes) [reply drape/vulkan/vulkan_base_context.cpp]: bool operator calls `has_value()` method

File: drape_frontend/frontend_renderer.cpp (hunk 2 above)
Comment (mpimenov) [reply drape_frontend/frontend_renderer.cpp]: Why is it better?
I preferred to leave has_value in the places where the original authors used is_initialized (boost::optional has operator bool too).

Comment (gmoryes) [reply drape_frontend/frontend_renderer.cpp]: okey

**Comment by maksimandrianov** [comment]
Кажется, кое что забыл в jni поправить
```
  ../../../../jni/com/mapswithme/maps/Framework.cpp:2003:51: error: no member named 'get' in 'std::__ndk1::optional<m2::Point<double> >'
    auto const latLon = mercator::ToLatLon(position.get());
                                           ~~~~~~~~ ^
  ../../../../jni/com/mapswithme/maps/Framework.cpp:2114:14: error: no member named 'is_initialized' in 'std::__ndk1::optional<eye::Tip::Type>'
    return tip.is_initialized() ? static_cast<jint>(tip.get()) : kUndefinedTip;
           ~~~ ^
  ../../../../jni/com/mapswithme/maps/Framework.cpp:2114:55: error: no member named 'get' in 'std::__ndk1::optional<eye::Tip::Type>'
    return tip.is_initialized() ? static_cast<jint>(tip.get()) : kUndefinedTip;
```

**Comment by mpimenov** [comment]
JTALL

**Comment by mpimenov** [comment]
JTALL

**Comment by mpimenov** [comment]
Only iOS build failed:
```
fatal error: error in backend: IO failure on output stream: No space left on device
Command CompileC failed with a nonzero exit code
```

**Pull request merged**